driver-ui/src/main/java/eu/tsystems/mms/tic/testframework/webdrivermanager/WebDriverSessionsManager.java

```diff
@@ -357,6 +360,26 @@ public static EventFiringWebDriver getWebDriver(final WebDriverRequest webDriver
         session already exists?
          */
         if (existingWebDriver != null) {
+            // Link sessionContext to methodContext if not exist
+            // e.g. Session was created in setup method and reused in test method
+            executionContextController.getCurrentSessionContext()
```

Comment:
This looks like a workaround (from hell) for the deduplication in `MethodContext#addSessionContext` being broken. (`SessionContext` does not seem to implement an `equals` - which it probably _really should_, if it has an `id`, so `this.sessionContexts.contains(sessionContext)` does reference comparison as per `Object#equals`.)

... Considering the... unprecedented... readability of this particular workaround, mayhap we should... _not_ workaround?

If need be, here is how I understood the code:
```java
executionContextController.getCurrentSessionContext().ifPresent(currentSessionContext -> executionContextController.getCurrentMethodContext()
	.filter(methodContext -> methodContext.readSessionContexts()
		.map(SessionContext::getRemoteSessionId)
		.filter(Optional::isPresent) // Optional.empty().equals(Optional.empty()) -> true
		.noneMatch(remoteSessionId -> currentSessionContext.getRemoteSessionId().equals(remoteSessionId)))
	.ifPresent(methodContext -> methodContext.addSessionContext(currentSessionContext)));
```

or possibly (closer to the original)

```java
	executionContextController.getCurrentSessionContext().ifPresent(currentSessionContext ->
		executionContextController.getCurrentMethodContext().ifPresent(currentMethodContext ->
			linkSessionContextToMethodContextIfNotExist(currentSessionContext, currentMethodContext)));

[...]

/** Link sessionContext to methodContext if not exist, e.g. if Session was created in setup method and reused in test method */
private static void linkSessionContextToMethodContextIfNotExist(final SessionContext sessionContext, final MethodContext methodContext) {
	final boolean sessionNotAmongMethodRemoteSessions = methodContext.readSessionContexts()
		.map(SessionContext::getRemoteSessionId)
		.filter(Optional::isPresent) // Optional.empty().equals(Optional.empty()) -> true
		.noneMatch(remoteSessionId -> sessionContext.getRemoteSessionId().equals(remoteSessionId));
	if (sessionNotAmongMethodRemoteSessions) {
		methodContext.addSessionContext(sessionContext);
	}
}
```

... If we cannot or should not fix `MethodContext#addSessionContext`, mayhap consider one of these forms (if I got them right, that is). Or write an entirely new one. Mayhap adding a new method to MethodContext (and deprecating the old one?) could be another option.

In any case, I must with all my heart utter my rejection of this particular block of... something.

Reply:
I've chosen your second proposal. Thanks for it!

Reply:
Shouldn't we fix `addSessionContext` instead? Overnight I got the idea that mayhap it does work after all, because `MethodContext#sessionContexts` might be custom Collection - but it's a default `java.util.LinkedList`. So the method _is_ broken and thence we deduplicate _twice_ - that cannot be healthy.

... In any case, as I proposed these as sufficient, I'll checkmark now.

Reply:
`addSessionContext()` already has a check about existing sessions. Therefor I removed the extra handling.